### ISSUE TITLE
N+1 issue when listing attachement and variants

### ISSUE THREAD

**Issue by paul-bonnel-fr** [rails/rails#46770]
### Steps to reproduce

```
  config.active_storage.service = :amazon
  config.active_storage.track_variants = true
```

```
    object.users.with_attached_thumbnail.each do |item|
      if item.thumbnail.present?
        puts item.thumbnail.variant(resize: "196x196^").processed.key
      end
    end

```

### Expected behavior
This code should only perform 1 SQL request. 

### Actual behavior
This code perform 1 SQL request et 3 others SQL request per thumbnail. 

```
 SELECT "active_storage_variant_records".* FROM "active_storage_variant_records" WHERE "active_storage_variant_records"."blob_id" = $1 AND "active_storage_variant_records"."variation_digest" = $2 LIMIT $3  [["blob_id", 52291], ["variation_digest", "IUN7MbalQqmLKbMoWq8RDR0hgdI="], ["LIMIT", 1]]

ctiveStorage::Attachment Load (0.2ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = $1 AND "active_storage_attachments"."record_type" = $2 AND "active_storage_attachments"."name" = $3 LIMIT $4  [["record_id", 20412], ["record_type", "ActiveStorage::VariantRecord"], ["name", "image"], ["LIMIT", 1]]

ActiveStorage::Blob Load (0.3ms)  SELECT "active_storage_blobs".* FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = $1 LIMIT $2  [["id", 52292], ["LIMIT", 1]]

```

### System configuration
**Rails version**:  7.0.4

**Ruby version**: 3.1.2

**Comment by nvasilevski** [comment]
Similar question came up in Discord recently and we could fix it by adding a condition based on `track_variants`
https://github.com/rails/rails/blob/41c2c26dc6ff03d35e11400ae767a7e518a7c065/activestorage/lib/active_storage/attached/model.rb#L73 to the scope generation logic to also preload variants if `track_variants` is `true`

Basically same as we do for the `has_many` association
https://github.com/rails/rails/blob/41c2c26dc6ff03d35e11400ae767a7e518a7c065/activestorage/lib/active_storage/attached/model.rb#L191-L195

But until it fixed, there is an option to explicitly use `includes` instead of the generated scope as shown in the code above to prevent n+1 query.

**Comment by ghiculescu** [comment]
Can you check if using the `main` branch fixes your issue? The 7.0.4 release does not include https://github.com/rails/rails/pull/42981 but I think that's what you're looking for here.

**Comment by paul-bonnel-fr** [comment]
Thank you very much guys for your anwser. I will let you know shorty if the main branch fix this issue. Cheers.

**Comment by paul-bonnel-fr** [comment]
Here is what I did : 
I've added remplace the line 
```gem 'rails',                                '7.0.4'```
with this one :
```gem 'rails',                                github: 'rails/rails'```
then : 
```bundle update```
But I still experience the n+1 issue with the exact same code in my first message. 

Let me know if you want me to try something else to help.

**Comment by nvasilevski** [comment]
Created a reproduction script
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", github: "rails/rails"
  gem "sqlite3"
  gem "debug"
end

require "debug"
require "active_record/railtie"
require "active_storage/engine"
require "tmpdir"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.eager_load = false
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  config.active_storage.service = :local
  config.active_storage.service_configurations = {
    local: {
      root: Dir.tmpdir,
      service: "Disk"
    }
  }
  config.active_storage.track_variants = true
end

ENV["DATABASE_URL"] = "sqlite3::memory:"

Rails.application.initialize!

require ActiveStorage::Engine.root.join("db/migrate/20170806125915_create_active_storage_tables.rb").to_s

ActiveRecord::Schema.define do
  CreateActiveStorageTables.new.change

  create_table :users, force: true
end

class User < ActiveRecord::Base
  has_one_attached :avatar do |attachable|
    attachable.variant :thumb, resize_to_limit: [100, 100]
  end
end

require "minitest/autorun"

class BugTest < Minitest::Test
  def test_upload_and_download
    User.create!(
      avatar: {
        content_type: "image/png",
        filename: "dummy1.png",
        io: ::StringIO.new("dummy"),
      }
    )

    User.create!(
      avatar: {
        content_type: "image/png",
        filename: "dummy2.png",
        io: ::StringIO.new("dummy2"),
      }
    )

    User.all.map { |u| u.avatar.variant(:thumb).url }

    current_sql = capture_sql do
      scope = User.with_attached_avatar
      scope.map { |u| u.avatar.variant(:thumb).url  }
    end

    includes_sql = capture_sql do
      scope = User.includes(avatar_attachment: {blob: { variant_records: { image_attachment: :blob } }})
      scope.map { |u| u.avatar.variant(:thumb).url  }
    end

    assert_equal(4, includes_sql.size) # User Load, ActiveStorage::Attachment Load, ActiveStorage::Blob Load,  ActiveStorage::VariantRecord Load
    assert_equal(4, current_sql.size) #  fails; same as above but there are two ActiveStorage::VariantRecord loads for each user
  end

  def capture_sql
    ActiveRecord::Base.connection.materialize_transactions
    SQLCounter.clear_log
    yield
    SQLCounter.log.dup
  end

  class SQLCounter
    class << self
      attr_accessor :ignored_sql, :log, :log_all
      def clear_log; self.log = []; self.log_all = []; end
    end

    clear_log

    def call(name, start, finish, message_id, values)
      return if values[:cached]

      sql = values[:sql]
      self.class.log_all << sql
      self.class.log << sql unless ["SCHEMA", "TRANSACTION"].include? values[:name]
    end
  end

  ActiveSupport::Notifications.subscribe("sql.active_record", SQLCounter.new)
end
```

The https://github.com/rails/rails/pull/42981 PR mentioned above fixes the issue for the `has_many` associations but the same fix should be applicable for the `has_one_attached` association

**Comment by cmrd-senya** [comment]
Hello!

I can see that the fix for this issue is merge to the `main` branch, but it's still absent from `7-0-stable` and the issue is present on the latest Rails 7 release, currently 7.0.8.

Is it a mistake or it was done so on purpose?

In what Rails release may we expect the fix to be included?

**Comment by skipkayhil** [comment]
> Is it a mistake or it was done so on purpose?
> 
> In what Rails release may we expect the fix to be included?

I think its an oversight so I opened a PR for `7-0-stable`: #49231. If that is merged it will be released in 7.0.9

Otherwise is will definitely be released with 7.1
